vm/examples/tic-tac-toe/app/src/game_manager.rs

```
@@ -0,0 +1,197 @@
+
+mod settings {
+    pub const PLAYERS_MAX_COUNT: usize = 1024;
+    pub const GAMES_MAX_COUNT: usize = 1024;
+}
+
+pub struct GameManager {
+    players: ArrayDeque<[Rc<RefCell<Player>>; settings::PLAYERS_MAX_COUNT], Wrapping>,
+    games: ArrayDeque<[Rc<RefCell<Game>>; settings::GAMES_MAX_COUNT], Wrapping>,
+    // TODO: String key should be replaced with Cow<'a, str>. After that signatures of all public
+    // functions also should be changed similar to https://jwilm.io/blog/from-str-to-cow/.
+    players_by_name: HashMap<String, Weak<RefCell<Player>>>,
+}
+
+impl GameManager {
+    pub fn new() -> Self {
+        GameManager {
+            players: ArrayDeque::new(),
+            games: ArrayDeque::new(),
+            players_by_name: HashMap::new(),
+        }
+    }
+
+    /// Marks an empty position on the board by user's tile type. Returns MoveResponse structure
+    /// as a serde_json Value.
+    pub fn make_move(
+        &self,
+        player_name: String,
+        player_sign: String,
+        coords: (usize, usize),
+    ) -> AppResult<Value> {
+        let game = self.get_player_game(&player_name, &player_sign)?;
+        let mut game = game.borrow_mut();
+        let game_move = GameMove::new(coords.0, coords.1)
+            .ok_or_else(|| format!("Invalid coordinates: x = {} y = {}", coords.0, coords.1))?;
+
+        let response = match game.player_move(game_move)? {
+            Some(app_move) => {
+                // checks did the app win in this turn?
+                let winner = match game.get_winner() {
+                    Some(winner) => winner.to_string(),
+                    None => "None".to_owned(),
+                };
+                MoveResponse {
+                    player_name,
+                    winner,
+                    coords: (app_move.x, app_move.y),
+                }
+            }
+            // none means a win of the player or a draw
+            None => MoveResponse {
```

Maybe return an error here? It's an error to make a move after game is ended anyways.

---

The really condition when error should be return is only here:
`game.player_move(game_move)?`
`player_move` checks has game been already finished. If so it return a error.

`None` could mean both a win of the player or a draw. These states can be reached after player move.